---
## Clean up

No Vector Search resources to clean up — the embeddings are stored in a Delta table (`main.default.ultrafeedback_embeddings`) which persists across assignments at no additional cost.


---
## 1. The human-in-the-loop *(Required)*

Automated judges are scalable but imperfect. They can miss domain nuances, apply criteria inconsistently, or disagree with expert opinions. **Human feedback** addresses this:

- A domain expert reviews agent outputs and rates them (good/bad, with rationale)
- The `align()` function uses this feedback to optimize the judge's prompt, making it better match human standards
- The aligned judge can then evaluate at scale, acting as a proxy for the human expert

This notebook aligns your custom judge to human feedback. The required path uses **SIMBA**, which revises the judge instructions based on examples where the current judge agrees or disagrees with your ratings.

### Requirements for alignment
- **Use the custom judge you created and registered in Assignment 5** (Section 8). In Assignment 5 you call `.register()` on your judge so it is stored in the experiment; here you load it with `get_scorer()`. Same name so your human feedback and `align()` target the same criteria.
- A judge created with `make_judge()` and registered (template-based)
- At least **10 traces** with human feedback (this notebook generates 12)
- The feedback assessment name must **exactly match** the judge name

---
## 2. Install dependencies *(Required)*

In [0]:
%pip install --upgrade "mlflow>=3.9" databricks-langchain unitycatalog-ai[databricks] numpy databricks-agents backoff "dspy>=3.1.0"
dbutils.library.restartPython()


*** WARNING: max output size exceeded, skipping output. ***

ollecting markdown-it-py>=2.2.0 (from rich>=12.3.0->typer->huggingface-hub<2.0,>=0.16.4->tokenizers->litellm<2,>=1.80.11->databricks-agents)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.2/10.2 MB 131.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 104.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 78.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 125.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.2/14.2 MB 131.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 829.5/829.5 kB 35.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 15.5/15.5 MB 148.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 54.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.2/12.2 MB 144.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 89.9 MB/s eta 0:00:00
   ━━━━━━

---
## 3. Recreate agent and custom judge *(Required)*

Load the agent from Assignment 4's `agent.py` (same approach as Assignment 5) and **load the custom judge you registered in Assignment 5** (Section 8) via `get_scorer()`. The judge you align in this assignment must be that same judge so your human feedback and `align()` apply to it. The embeddings table is verified first; then the agent is loaded via a wrapper so `agent_executor.invoke({"input": question})` works for trace generation.

### 3a. Verify embeddings table

The embeddings table (`main.default.ultrafeedback_embeddings`) was created in Assignment 3. Verify it exists and has the expected data.


In [0]:
# Verify the embeddings table from Assignment 3 exists
emb_df = spark.table("main.default.ultrafeedback_embeddings")
print(f"Embeddings table has {emb_df.count()} rows.")
print(f"Columns: {emb_df.columns}")
display(emb_df.select("id", "instruction", "source").limit(3))


Embeddings table has 100 rows.
Columns: ['id', 'instruction', 'source', 'embedding']


id,instruction,source
0,"Can you write a shell command that can print the SHA-256 hash of a file named ""document.pdf""? And also, can you write a shell command that can rename all the files in a directory with a "".txt"" extension to have a "".md"" extension instead?",evol_instruct
1,"Can you provide an example of a well-written constructive feedback that includes specific and actionable suggestions on how to improve an individual's performance in a particular task or project, while also acknowledging their strengths? Please write it down in [Mark down] format, making sure to provide detailed reasoning behind each suggestion. Furthermore, elaborate on the crucial role constructive feedback plays in fostering a positive and collaborative team or work environment, as it not only helps individuals improve but also contributes to the overall success of the team and the organization.",evol_instruct
2,Classify the given element as an acid or a base. Carbonic acid,evol_instruct


### 3b. Load agent from agent.py

Same approach as Assignment 5: load the Assignment 4 agent from `agent.py` (ensure it is in the same directory as this notebook or on the Python path). The wrapper exposes `agent_executor.invoke({"input": question})` returning `{"output": "..."}` for trace generation below.


In [0]:
import mlflow

# Enable tracing for agent invocations (traces used in Section 4–5)
mlflow.langchain.autolog()

# Shared experiment across Assignments 4, 5, 6
EXPERIMENT_NAME = "/Users/" + spark.sql("SELECT current_user()").first()[0] + "/aai594_ultrafeedback_expert"
mlflow.set_experiment(EXPERIMENT_NAME)

# Load the agent from Assignment 4's agent.py (same as Assignment 5)
import agent
AGENT = agent.ToolCallingAgent(llm_endpoint=agent.LLM_ENDPOINT_NAME, tools=agent.TOOL_INFOS)


def _response_to_text(resp):
    """Extract final answer text from ResponsesAgentResponse."""
    text_parts = []
    for item in resp.output:
        c = getattr(item, "content", None)
        if c is None:
            continue
        for part in (c if isinstance(c, list) else [c]):
            if isinstance(part, dict) and "text" in part:
                text_parts.append(part["text"])
            elif hasattr(part, "text"):
                text_parts.append(part.text)
    return "\n".join(text_parts) if text_parts else str(resp.output)[:500]


# Wrapper so trace generation can use agent_executor.invoke({"input": question}) -> {"output": "..."}
class AgentExecutorWrapper:
    def invoke(self, inputs):
        resp = AGENT.predict({
            "input": [{"role": "user", "content": inputs["input"]}],
            "custom_inputs": {"session_id": "aai594-traces"},
        })
        return {"output": _response_to_text(resp)}


agent_executor = AgentExecutorWrapper()



/databricks/python/lib/python3.11/site-packages/databricks/connect/session.py:426: UserWarning: Ignoring the default notebook Spark session and creating a new Spark Connect session. To use the default notebook Spark session, use DatabricksSession.builder.getOrCreate() with no additional parameters.
  warnings.warn(new_notebook_session_msg)


In [0]:
import mlflow
from mlflow.genai import get_scorer

# Load the custom judge you created and registered in Assignment 5 (Section 8)
# Use the same JUDGE_NAME you used when you registered the scorer in Assignment 5
JUDGE_NAME = "tool_usage_quality"  # Use your judge name from Assignment 5
EXPERIMENT_NAME = "/Users/" + spark.sql("SELECT current_user()").first()[0] + "/aai594_ultrafeedback_expert"

# Get the experiment ID (same experiment as Assignment 5) and load the scorer
experiment_id = mlflow.get_experiment_by_name(EXPERIMENT_NAME).experiment_id
custom_judge= get_scorer(name=JUDGE_NAME, experiment_id=experiment_id)

print(f"Loaded judge '{JUDGE_NAME}' from experiment.")

Loaded judge 'tool_usage_quality' from experiment.


In [0]:
# Quick test
resp = agent_executor.invoke({"input": "How many rows come from evol_instruct?"})
print("Agent test:", resp["output"][:200])

/local_disk0/.ephemeral_nfs/envs/pythonEnv-1ee4ee83-4b65-430c-9e0b-7d1dedfdb124/lib/python3.11/site-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected `str` - serialized value may not be as expected [field_name='content', input_value=[{'type': 'reasoning', 's...tion I can help with!'}], input_type=list])
  return self.__pydantic_serializer__.to_python(


Agent test: I’m sorry, but I don’t have a tool that can retrieve the number of rows originating from the **evol_instruct** source in the UltraFeedback dataset. If you have access to the raw dataset you could coun


Trace(trace_id=tr-651e4569816cabc97d0b7f4e39d963be)

### 3c. Load the custom judge you registered in Assignment 5

**Load the custom judge you created and registered in Assignment 5** (Section 8). In Assignment 5 you called `.register()` on your judge so it is stored in the experiment. Here we load it with `get_scorer(name=JUDGE_NAME, experiment_id=...)`. Use the **same judge name** you used in Assignment 5 (e.g. `tool_usage_quality` or your own name). That judge is the one you will log feedback for (Section 5) and align (Section 6). The feedback assessment name when you log feedback must **exactly match** this judge's name.

**Docs:** [Versioning Scorers](https://mlflow.org/docs/latest/genai/eval-monitor/scorers/versioning/)


## 4. Generate 12 traces *(Required)*

Run the agent on 12 diverse questions using agent_executor.invoke({"input": question}) (same interface as Assignment 5). Each invocation is logged as a trace in MLflow. Traces capture the full execution: inputs, outputs, tool calls, and intermediate steps.

The questions cover different tool types and edge cases to give you varied outputs to review.

In [0]:
# 12 diverse questions for trace generation (target tools from Assignment 3)
TRACE_QUESTIONS = [
    # all_model_combinations
    "Which model combinations have the biggest disparity vs. chosen and rejected?",
    "Which model combinations have the most records? Show counts and average ratings.",
    # compare_models
    "How does gpt-4 compare to llama-2-7b-chat in the dataset?",
    "Head to head: gpt-4 vs wizardlm-7b. Which model was chosen more often?",
    # analyze_instruction
    "Analyze the complexity of: What is the weather today?",
    "Analyze this instruction: Design a comprehensive machine learning pipeline that ingests real-time streaming data, performs feature engineering, and deploys the best model to a REST API.",
    # Complex questions
    "Analyze this instruction: How many windows are there in the Empire State Building? Are there similar questions on this topic?",
    "Find instructions similar to wild rice, and which model is the best?",
    # search_similar_instructions
    "Find instructions in the dataset similar to 'python programming'. Show a few examples.",
    "Search for instructions about quantum computing.",
    "What instructions in the dataset mention health benefits?",
    # Multi-step or edge cases
    "Find instructions about writing code, then analyze the complexity of one of them.",
]

# Run all queries and store results with the official MLflow tagging
trace_results = []
import mlflow
for i, question in enumerate(TRACE_QUESTIONS):
    print(f"\n--- Trace {i+1}/12: {question[:60]}... ---")
    try:        
        response = agent_executor.invoke({"input": question})
        output = response["output"]
        print(f"Output: {output[:150]}...")
        trace_results.append({
            "question": question,
            "output": output,
            "error": None
        })
    except Exception as e:
        print(f"Error: {e}")
        trace_results.append({
            "question": question,
            "output": None,
            "error": str(e)
        })

print(f"\n=== Generated {len(trace_results)} traces ===")


--- Trace 1/12: Which model combinations have the biggest disparity vs. chos... ---


/local_disk0/.ephemeral_nfs/envs/pythonEnv-1ee4ee83-4b65-430c-9e0b-7d1dedfdb124/lib/python3.11/site-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected `str` - serialized value may not be as expected [field_name='content', input_value=[{'type': 'reasoning', 's...n drill down further!'}], input_type=list])
  return self.__pydantic_serializer__.to_python(


Output: **Answer – Model pairs with the largest gap between the average “chosen” rating and the average “rejected” rating**

The table returned by `main__defa...

--- Trace 2/12: Which model combinations have the most records? Show counts ... ---


/local_disk0/.ephemeral_nfs/envs/pythonEnv-1ee4ee83-4b65-430c-9e0b-7d1dedfdb124/lib/python3.11/site-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected `str` - serialized value may not be as expected [field_name='content', input_value=[{'type': 'reasoning', 's...e “rejected” one.'}], input_type=list])
  return self.__pydantic_serializer__.to_python(


Output: **Top model‑pairings by number of preference records**  
*(Data extracted with `main__default__all_model_combinations`)*  

| Rank | Chosen model | Re...

--- Trace 3/12: How does gpt-4 compare to llama-2-7b-chat in the dataset?... ---


/local_disk0/.ephemeral_nfs/envs/pythonEnv-1ee4ee83-4b65-430c-9e0b-7d1dedfdb124/lib/python3.11/site-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected `str` - serialized value may not be as expected [field_name='content', input_value=[{'type': 'reasoning', 's...Llama‑2 chat model.'}], input_type=list])
  return self.__pydantic_serializer__.to_python(


Output: **Comparison of GPT‑4 vs. Llama‑2‑7B‑Chat in the UltraFeedback dataset**

| Model (winner) | Times chosen over the other | Avg. chosen rating | Avg. r...

--- Trace 4/12: Head to head: gpt-4 vs wizardlm-7b. Which model was chosen m... ---


/local_disk0/.ephemeral_nfs/envs/pythonEnv-1ee4ee83-4b65-430c-9e0b-7d1dedfdb124/lib/python3.11/site-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected `str` - serialized value may not be as expected [field_name='content', input_value=[{'type': 'reasoning', 's...d over WizardLM‑7B.'}], input_type=list])
  return self.__pydantic_serializer__.to_python(


Output: **Result:** In the UltraFeedback dataset, **GPT‑4 was chosen far more often than WizardLM‑7B**.

**Details from the comparison**

| Model | Times Chos...

--- Trace 5/12: Analyze the complexity of: What is the weather today?... ---


/local_disk0/.ephemeral_nfs/envs/pythonEnv-1ee4ee83-4b65-430c-9e0b-7d1dedfdb124/lib/python3.11/site-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected `str` - serialized value may not be as expected [field_name='content', input_value=[{'type': 'reasoning', 's...stand and respond to.'}], input_type=list])
  return self.__pydantic_serializer__.to_python(
/local_disk0/.ephemeral_nfs/envs/pythonEnv-1ee4ee83-4b65-430c-9e0b-7d1dedfdb124/lib/python3.11/site-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected `str` - serialized value may not be as expected [field_name='content', input_value=[{'type': 'reasoning', 's...stand and respond to.'}], input_type=list])
  return self.__pydantic_serializer__.to_python(


Output: **Complexity Analysis of the instruction**  
*Instruction:* **“What is the weather today?”**

| Metric | Value |
|--------|-------|
| **Word count** |...

--- Trace 6/12: Analyze this instruction: Design a comprehensive machine lea... ---


/local_disk0/.ephemeral_nfs/envs/pythonEnv-1ee4ee83-4b65-430c-9e0b-7d1dedfdb124/lib/python3.11/site-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected `str` - serialized value may not be as expected [field_name='content', input_value=[{'type': 'reasoning', 's...in a single sentence.'}], input_type=list])
  return self.__pydantic_serializer__.to_python(


Output: **Analysis of the instruction**

| Metric | Value |
|--------|-------|
| **Word count** | 23 |
| **Sentence count** | 1 |
| **Average word length** | ...

--- Trace 7/12: Analyze this instruction: How many windows are there in the ... ---


/local_disk0/.ephemeral_nfs/envs/pythonEnv-1ee4ee83-4b65-430c-9e0b-7d1dedfdb124/lib/python3.11/site-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected `str` - serialized value may not be as expected [field_name='content', input_value=[{'type': 'reasoning', 's...Building’s windows.'}], input_type=list])
  return self.__pydantic_serializer__.to_python(


Output: **Instruction analysis**

| Metric | Value |
|--------|-------|
| **Word count** | 17 |
| **Sentence count** | 2 |
| **Average word length** | 4.9 cha...

--- Trace 8/12: Find instructions similar to wild rice, and which model is t... ---


/local_disk0/.ephemeral_nfs/envs/pythonEnv-1ee4ee83-4b65-430c-9e0b-7d1dedfdb124/lib/python3.11/site-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected `str` - serialized value may not be as expected [field_name='content', input_value=[{'type': 'reasoning', 's...give you the results.'}], input_type=list])
  return self.__pydantic_serializer__.to_python(


Output: I ran a similarity search for the phrase **“wild rice”** using the UltraFeedback dataset (via `search_similar_instruct`). The tool returned **no match...

--- Trace 9/12: Find instructions in the dataset similar to 'python programm... ---


/local_disk0/.ephemeral_nfs/envs/pythonEnv-1ee4ee83-4b65-430c-9e0b-7d1dedfdb124/lib/python3.11/site-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected `str` - serialized value may not be as expected [field_name='content', input_value=[{'type': 'reasoning', 's...feature explanations.'}], input_type=list])
  return self.__pydantic_serializer__.to_python(


Output: **Using `search_similar_instruct` for the query “python programming” (top 3 results)**  

| # | Matching instruction (truncated) | Source |
|---|-----...

--- Trace 10/12: Search for instructions about quantum computing.... ---


/local_disk0/.ephemeral_nfs/envs/pythonEnv-1ee4ee83-4b65-430c-9e0b-7d1dedfdb124/lib/python3.11/site-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected `str` - serialized value may not be as expected [field_name='content', input_value=[{'type': 'reasoning', 's...e what it returned."}]}], input_type=list])
  return self.__pydantic_serializer__.to_python(


Output: Max iterations reached. Stopping....

--- Trace 11/12: What instructions in the dataset mention health benefits?... ---


/local_disk0/.ephemeral_nfs/envs/pythonEnv-1ee4ee83-4b65-430c-9e0b-7d1dedfdb124/lib/python3.11/site-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected `str` - serialized value may not be as expected [field_name='content', input_value=[{'type': 'reasoning', 's...ee the tool output."}]}], input_type=list])
  return self.__pydantic_serializer__.to_python(


Output: Max iterations reached. Stopping....

--- Trace 12/12: Find instructions about writing code, then analyze the compl... ---
Output: Max iterations reached. Stopping....

=== Generated 12 traces ===


[Trace(trace_id=tr-31ce9b1b241651d75c58bbfed2f9d486), Trace(trace_id=tr-50ab56c27c5bd37bdc5e7865568baeb1), Trace(trace_id=tr-90dcdf2d3b094fe22501a50f46a8394a), Trace(trace_id=tr-acb60b9d5ca7fc96cd7c0bb0052ec2b0), Trace(trace_id=tr-63371cd809881c543e742cbb2055c411), Trace(trace_id=tr-b1862e9f5fe99c58d4eaf0e5c4b2728d), Trace(trace_id=tr-f770adc8dfb1df6d6ecea6adfb5d0ae3), Trace(trace_id=tr-f49455c85433d3dc365685efb5038823), Trace(trace_id=tr-bbd13e7961f737c09a8cc59807f5127c), Trace(trace_id=tr-82213554416cab8888036e056cc3db78)]

In [0]:
# Retrieve trace IDs from MLflow - get the most recent 12
# This is just here so you don't have to keep rerunning the trace creation if you come back to this after the notebook resets
traces = mlflow.search_traces(
    max_results=12,
    order_by=["timestamp_ms DESC"],
)

print(f"Found {len(traces)} recent traces in experiment.")


Found 12 recent traces in experiment.


In [0]:
# Assuming 'traces' already contains all relevant rows/columns
# If your 'inputs' column isn't named 'inputs' or 'response', rename as needed
answer_sheet_df = traces.rename(columns={'response': 'output'})  # MLflow usually expects 'output'
# If 'inputs' is nested, flatten as needed

result = mlflow.genai.evaluate(
    data=answer_sheet_df,
    scorers=[custom_judge] 
)

2026/03/14 18:10:30 INFO mlflow.models.evaluation.utils.trace: Auto tracing is temporarily enabled during the model evaluation for computing some metrics and debugging. To disable tracing, call `mlflow.autolog(disable=True)`.


Evaluating:   0%|          | 0/12 [Elapsed: 00:00, Remaining: ?] 

In [0]:
# Tag each trace for alignment - REQUIRED for later filtering
import mlflow

trace_ids = traces['trace_id'].values

for trace_id in trace_ids:
    mlflow.set_trace_tag(trace_id=trace_id, key="eval", value="align")

print(f"Tagged {len(traces)} traces with eval: align tag.")

com.databricks.backend.common.rpc.CommandCancelledException
	at com.databricks.spark.chauffeur.ExecContextState.cancel(ExecContextState.scala:434)
	at com.databricks.spark.chauffeur.ExecutionContextManagerV1.cancelExecution(ExecutionContextManagerV1.scala:466)
	at com.databricks.spark.chauffeur.ChauffeurState.$anonfun$process$1(ChauffeurState.scala:757)
	at com.databricks.logging.UsageLogging.$anonfun$recordOperation$1(UsageLogging.scala:510)
	at com.databricks.logging.UsageLogging.executeThunkAndCaptureResultTags$1(UsageLogging.scala:616)
	at com.databricks.logging.UsageLogging.$anonfun$recordOperationWithResultTags$4(UsageLogging.scala:643)
	at com.databricks.logging.AttributionContextTracing.$anonfun$withAttributionContext$1(AttributionContextTracing.scala:49)
	at com.databricks.logging.AttributionContext$.$anonfun$withValue$1(AttributionContext.scala:293)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
	at com.databricks.logging.AttributionContext$.withValue(Attr

---
## 5. Review traces and log feedback *(Required)*

Now you become the **domain expert**. For each trace:

1. **Review the trace** in the MLflow UI (Experiments > your experiment > Traces tab). Look at:
   - Did the agent call the right tool?
   - Was the response accurate?
   - Did the agent handle edge cases well?

2. **Log feedback** using the MLflow experiments UI.
   - Choose the **same judge name** and feedback type as your custom judge.
   - Provide a rating that reflects your real assessment of the trace.
   - Add a short rationale for each label so the alignment step has useful signal.
   - A good alignment set usually includes both positive and negative examples. If all 12 traces receive the same label, revisit your trace set or your labeling criteria.

**Important:** The assessment name must exactly match the judge name for `align()` to work. Use the same judge you recreated in Section 3c.

**Docs:** [Label during development](https://docs.databricks.com/aws/en/mlflow3/genai/human-feedback/dev-annotations) · [Feedback collection](https://mlflow.org/docs/latest/genai/assessments/feedback/)

---
## 6. Align the judge *(Required)*

The `align()` function improves **the custom judge you created in Assignment 5** (loaded in Section 3c) based on your human feedback. This notebook uses **SIMBAAlignmentOptimizer**, which iteratively proposes instruction updates and keeps the ones that best match your labels. This is the only required alignment path in this assignment.

**Requirements:**
- The judge you recreated in Section 3c (same as your Assignment 5 custom judge)
- At least 10 traces with feedback (we have 12), tagged with `eval='align'`
- Feedback assessment name matches the judge name
- MLflow 3.9+ for alignment optimizers

**Docs:** [Align judges with humans](https://docs.databricks.com/aws/en/mlflow3/genai/eval-monitor/align-judges) · [SIMBA](https://mlflow.org/docs/latest/genai/eval-monitor/scorers/llm-judge/simba/) · [MemAlign](https://mlflow.org/docs/latest/genai/eval-monitor/scorers/llm-judge/memalign/)

If you are curious about the alternative memory-based approach, read the MemAlign docs above. It is not part of the required assignment flow.


In [0]:
# Retrieve the labeled traces that will be used for alignment.
traces_for_alignment = mlflow.search_traces(
    filter_string="tag.eval = 'align'",
    max_results=12,
    return_type="list",
)

# Keep the original trace objects because alignment expects full trace metadata.
print(f"Found {len(traces_for_alignment)} traces with tag eval='align' in experiment.")

Found 12 traces with tag eval='align' in experiment.


In [0]:
import logging
import sys

logging.basicConfig(
    level=logging.WARNING,
    format="%(asctime)s %(levelname)s %(name)s: %(message)s",
    force=True,
    stream=sys.stdout,
)

# Keep only SIMBA-related logs. Switch any of these to DEBUG if you need deeper troubleshooting.
logging.getLogger("SIMBAAlignmentOptimizer").setLevel(logging.INFO)
logging.getLogger("mlflow.genai.judges.optimizers.simba").setLevel(logging.INFO)
logging.getLogger("dspy.teleprompt.simba").setLevel(logging.INFO)

# Silence noisy internals
for name in [
    "mlflow",
    "mlflow.genai.judges.optimizers.dspy_utils",
    "databricks.sdk",
    "httpcore",
    "httpx",
    "urllib3",
]:
    logging.getLogger(name).setLevel(logging.WARNING)


In [0]:
# Required path: align the judge with SIMBA.
# Expect this to take several minutes. If you hit rate limits, rerun the cell or reduce max_steps.
# Alignment may improve the judge, but it is not guaranteed to outperform your original prompt on every trace.

# Align the judge using traces with SME feedback
from mlflow.genai.judges import make_judge
from mlflow.genai.judges.optimizers import SIMBAAlignmentOptimizer
optimizer = SIMBAAlignmentOptimizer(
    model="databricks:/databricks-gpt-oss-120b",  # Optimizer model used to propose judge revisions.
    simba_kwargs={
        "max_steps": 2,  # Each step runs another optimization round. Lower is faster.
        "max_demos": 0,  # Rule-only optimization. Set >0 to let SIMBA add few-shot demos.
    },
)
aligned_judge = custom_judge.align(traces=traces_for_alignment, optimizer=optimizer)
print(f"Alignment complete for judge: {aligned_judge.name}")

2026-03-14 19:35:54,418 DEBUG SIMBAAlignmentOptimizer: Setting up DSPy context with model: databricks:/databricks-gpt-oss-120b
2026-03-14 19:35:54,420 DEBUG SIMBAAlignmentOptimizer: Created DSPy program with signature using judge's model
2026-03-14 19:35:54,423 INFO SIMBAAlignmentOptimizer: Preparing optimization with 12 examples from 12 traces
2026-03-14 19:35:54,424 DEBUG SIMBAAlignmentOptimizer: Starting DSPy optimization...


2026/03/14 19:35:54 INFO mlflow.genai.judges.optimizers.simba: Starting SIMBA optimization with 12 examples (set logging to DEBUG for detailed output)


Processed 6 / 60 examples:  10%|█         | 6/60 [00:31<06:31,  7.25s/it]

2026/03/14 19:36:51 WARNING dspy.teleprompt.simba_utils: Failed to invoke the judge via litellm: litellm.RateLimitError: DatabricksException - {"error_code":"REQUEST_LIMIT_EXCEEDED","message":"REQUEST_LIMIT_EXCEEDED: Exceeded workspace output tokens per second rate limit for databricks-gpt-oss-120b. Please use a provisioned throughput Foundation Model APIs endpoint for a higher rate limit."}


Processed 8 / 60 examples:  13%|█▎        | 8/60 [01:04<09:26, 10.90s/it]

2026/03/14 19:37:00 WARNING dspy.teleprompt.simba_utils: Failed to invoke the judge via litellm: litellm.RateLimitError: DatabricksException - {"error_code":"REQUEST_LIMIT_EXCEEDED","message":"REQUEST_LIMIT_EXCEEDED: Exceeded workspace QPS rate limit for databricks-gpt-oss-120b. Please use a provisioned throughput Foundation Model APIs endpoint for a higher rate limit."}


Processed 60 / 60 examples: 100%|██████████| 60/60 [02:30<00:00,  2.52s/it]


2026/03/14 19:38:31 INFO dspy.teleprompt.simba_utils: Advice for self: If the input `inputs` contains a user request that includes a factual question (e.g., asking for a numeric value, a date, a name, etc.) *in addition* to an instruction‑analysis or similarity‑search request, then you must:
1. Run the instruction‑analysis tool as before to compute word count, sentence count, complexity, etc., **but limit the complexity calculation to the first sentence/question only**.
2. Recognize that the factual question requires external knowledge. Invoke a knowledge‑retrieval tool (e.g., a web‑search / You.com MCP or any configured factual‑lookup tool) to obtain the answer.
3. Incorporate the retrieved factual answer into the final response.
4. After both steps, evaluate tool usage: 
   - If the factual answer was obtained and the similarity‑search results were used correctly, set `result = true`.
   - If any required tool (especially the knowledge‑retrieval tool for the factual part) was omitted

Processed 41 / 70 examples:  59%|█████▊    | 41/70 [00:36<01:23,  2.87s/it]

2026/03/14 19:39:56 WARNING dspy.teleprompt.simba_utils: Failed to invoke the judge via litellm: litellm.RateLimitError: DatabricksException - {"error_code":"REQUEST_LIMIT_EXCEEDED","message":"REQUEST_LIMIT_EXCEEDED: Exceeded workspace output tokens per second rate limit for databricks-gpt-oss-120b. Please use a provisioned throughput Foundation Model APIs endpoint for a higher rate limit."}


Processed 45 / 70 examples:  64%|██████▍   | 45/70 [01:21<02:33,  6.14s/it]

2026/03/14 19:40:03 WARNING dspy.teleprompt.simba_utils: Failed to invoke the judge via litellm: litellm.RateLimitError: DatabricksException - {"error_code":"REQUEST_LIMIT_EXCEEDED","message":"REQUEST_LIMIT_EXCEEDED: Exceeded workspace QPS rate limit for databricks-gpt-oss-120b. Please use a provisioned throughput Foundation Model APIs endpoint for a higher rate limit."}


Processed 60 / 60 examples: 100%|██████████| 60/60 [01:57<00:00,  1.96s/it]


2026/03/14 19:42:33 INFO dspy.teleprompt.simba_utils: Advice for self: If the input `inputs` contains a request that starts with "Analyze this instruction:" and the quoted instruction itself includes a factual question (e.g., "How many windows are there in the Empire State Building?"), then:
1. **Run the instruction‑analysis tool** (`analyze_instruction`) on the *first* sentence only. Compute word‑count, sentence‑count, average word length, complexity level, and the "is a question" flag for that sentence; ignore any follow‑up sentence that asks for additional actions.
2. **Detect any factual query** embedded in that first sentence. When such a query is present, **invoke a knowledge‑retrieval tool** (e.g., `search_web`, `search_knowledge_base`, or any configured external search tool) to obtain the factual answer.
3. **If the user also asks for similar instructions**, invoke the similarity‑search tool after step 1 (or concurrently) and report whether matches are found.
4. **Compose the f

Processed 47 / 70 examples:  67%|██████▋   | 47/70 [00:37<00:33,  1.47s/it]

2026/03/14 19:44:21 WARNING dspy.teleprompt.simba_utils: Failed to invoke the judge via litellm: litellm.RateLimitError: DatabricksException - {"error_code":"REQUEST_LIMIT_EXCEEDED","message":"REQUEST_LIMIT_EXCEEDED: Exceeded workspace output tokens per second rate limit for databricks-gpt-oss-120b. Please use a provisioned throughput Foundation Model APIs endpoint for a higher rate limit."}


Processed 58 / 70 examples:  83%|████████▎ | 58/70 [01:34<00:11,  1.02it/s]

2026/03/14 19:44:31 WARNING dspy.teleprompt.simba_utils: Failed to invoke the judge via litellm: litellm.RateLimitError: DatabricksException - {"error_code":"REQUEST_LIMIT_EXCEEDED","message":"REQUEST_LIMIT_EXCEEDED: Exceeded workspace QPS rate limit for databricks-gpt-oss-120b. Please use a provisioned throughput Foundation Model APIs endpoint for a higher rate limit."}


Processed 69 / 70 examples:  99%|█████████▊| 69/70 [01:40<00:00,  2.47it/s]

2026/03/14 19:44:39 WARNING dspy.teleprompt.simba_utils: Failed to invoke the judge via litellm: litellm.RateLimitError: DatabricksException - {"error_code":"REQUEST_LIMIT_EXCEEDED","message":"REQUEST_LIMIT_EXCEEDED: Exceeded workspace QPS rate limit for databricks-gpt-oss-120b. Please use a provisioned throughput Foundation Model APIs endpoint for a higher rate limit."}


Processed 36 / 36 examples: 100%|██████████| 36/36 [01:09<00:00,  1.94s/it]

2026/03/14 19:45:49 INFO mlflow.genai.judges.optimizers.simba: SIMBA optimization completed



2026-03-14 19:45:49,142 DEBUG SIMBAAlignmentOptimizer: DSPy optimization completed
Alignment complete for judge: tool_usage_quality


In [0]:
print('ORIGINAL JUDGE:\n')
print(custom_judge.instructions)
print('\n')
print("_"*100)
print("\nALIGNED JUDGE\n")
print(aligned_judge.instructions)

ORIGINAL JUDGE:

You are evaluating an AI agent's ability to use tools appropriately.

The agent has access to these tools (from Assignments 3–4):
- all_model_combinations: list model pairs and comparison data from the dataset
- analyze_instruction: instruction complexity metrics (word count, sentence count, etc.)
- compare_models: compare two models (e.g. win counts, average chosen/rejected ratings)
- search_similar_instructions: search the embeddings table for similar instructions by keyword
- Any other main.default.* UC functions you created in Assignment 3
- You.com MCP (if configured): live web search

User question: {{inputs}}
Agent response: {{outputs}}

Evaluate the agent's tool usage:
1. Did the agent call the appropriate tool(s) for this question?
2. If the question didn't need a tool, did the agent correctly avoid using one?
3. Did the agent use the tool results correctly in its response?

Return YES if tool usage was appropriate, NO if it was not.
Explain your reasoning.




### Compare the original and aligned judges

Rerun evaluation on the same traces so you can compare how the **original** and **aligned** judges score identical outputs.

In [0]:
print("\nRunning aligned judge for comparison...")
traces_for_eval = mlflow.search_traces(
    filter_string="tag.eval = 'align'",
    max_results=12,
)
answer_sheet_df_align = traces_for_eval.rename(columns={'response': 'output'})  # MLflow usually expects 'output'

aligned_results = mlflow.genai.evaluate(
    data=answer_sheet_df_align,
    scorers=[aligned_judge]
)



Running aligned judge...


Evaluating:   0%|          | 0/12 [Elapsed: 00:00, Remaining: ?] 

---
## 7. Reflection *(Required)*

### 7a. How did alignment change the judge? Did it reflect your manually provided feedback?

*Compare the original and aligned judge prompts. What changed? Did the aligned judge add new criteria, change the emphasis, or use different language? Did the new evaluation job lead to different results?*

**[Your answer here]**

### 7b. What does this teach about human-in-the-loop evaluation?

*Reflect on the entire arc: subjective reviews (Week 4) → automated judges (Week 5) → human feedback → alignment (this week). How does this connect to the principle of humans steering agents?*

**[Your answer here]**

*The required assignment ends after you answer these two reflection prompts. Everything below is optional enrichment.*

---
## Optional Extension: optimize_anything and agent skills

*This section is not required for submission and is not part of the grading rubric.*

**What it is:** [GEPA](https://gepa-ai.github.io/gepa/)'s **optimize_anything** API optimizes *any* text artifact — prompts, code, configs, or **agent skills** — using an evaluator and LLM-driven reflection. You declare what to optimize and how to measure it; the system handles the search (single-task, multi-task, or generalization). Unlike hand-written mutation prompts, you provide a **seed** (or a natural-language objective), an **evaluator** that scores candidates and can return **Actionable Side Information (ASI)** (e.g. error messages, execution traces), and optionally a **dataset** and **valset** so the optimized artifact generalizes to unseen examples.

**Why agent skills matter:** Agent **skills** are natural-language instructions and best practices that help an agent perform well on a specific codebase or domain. They are text artifacts: if you can run the agent on real tasks and score outcomes, you can optimize them. In practice, GEPA-optimized skills have been shown to push coding agents to near-perfect task completion while being **47% faster** and to **transfer across models** (e.g. skills learned for one repo improving another). So learning to generate or refine agent skills with optimize_anything is a natural next step after judge alignment: you go from "align a judge to human feedback" to "optimize the instructions the agent follows so it succeeds on more tasks."

**Giving it an objective:** You don't have to provide a starting artifact. You can call `optimize_anything` with an **objective** — a natural-language description of what you want — and no `seed_candidate`. The reflection LLM then bootstraps the first candidate from that description ("seedless" mode). For example: *"Generate agent skills (natural-language instructions) that help the UltraFeedback Expert use its tools appropriately on real user questions and generalize to unseen tasks."* The evaluator still runs your agent on a dataset and returns scores (and ASI); the optimizer proposes and refines candidates until they meet your objective. This is especially useful when the solution space is large or you're not sure what a good seed looks like.

**How to get started:**

1. **Install:** `pip install gepa` (must be >=0.1.0)
2. **Define an evaluator** that takes a candidate (e.g. a skill document or prompt string), runs your agent on a set of tasks, and returns a score. The evaluator **produces** ASI: either return a `(score, side_info)` tuple or call `oa.log()` inside the evaluator so the optimizer has diagnostics (errors, traces) to guide reflection.
3. **Call** `optimize_anything(evaluator=evaluate, objective="...", dataset=..., valset=...)` with an **objective** and no seed to go seedless, or pass `seed_candidate=...` if you have a starting draft. Use **generalization** mode (dataset + valset) so the optimized skill transfers to unseen tasks.

For full examples (including the agent-skills case study), code walkthroughs, and the three optimization modes, use the links below.

**Where to go next:**

- **Blog — optimize_anything intro (agent skills, cloud algorithms, ARC-AGI, etc.):** [Introducing optimize_anything](https://gepa-ai.github.io/gepa/blog/2026/02/18/introducing-optimize-anything/)
- **GEPA site (quick start, adapters, tutorials):** [gepa-ai.github.io/gepa](https://gepa-ai.github.io/gepa/)
- **Repo and API reference:** [github.com/gepa-ai/gepa](https://github.com/gepa-ai/gepa)



---
## Lab complete

### Required (Sections 1-7)
- [ ] **Section 3:** Agent and custom judge recreated. Vector Search working.
- [ ] **Section 4:** 12 traces generated and visible in MLflow Experiments.
- [ ] **Section 5:** Human feedback logged for all 12 traces with honest ratings and rationales.
- [ ] **Section 6:** SIMBA alignment ran successfully. Original and aligned judge were compared on the same traces.
- [ ] **Section 7:** Both reflection questions answered.
- [ ] **Cleanup note:** Reviewed the cleanup note at the top of the notebook. No additional cleanup is required for this lab.

**Submit:** Your executed notebook (`.ipynb` with all outputs) and the completed `SUBMISSION_6.md`.

*You have completed the required assignment flow. The optimize_anything section above is optional enrichment and is not required for submission.*